HF Buckets integration to Trainer (#46386)
huggingface/transformers: docs/source/en/trainer_recipes.md

When resuming, [`Trainer`] restores the optimizer state, scheduler state, and RNG state.

Checkpoint resuming requires optimizer and scheduler state files in the checkpoint directory. If those files are missing (for example, when `save_only_model=True`), the optimizer restarts from scratch.

### Checkpointing to a bucket

> Set `push_to_bucket=True` to sync checkpoints to a [Hugging Face bucket](https://huggingface.co/docs/huggingface_hub/guides/buckets) instead of (or alongside) keeping them only on local disk. A bucket is mutable, Xet-backed object storage with no git history — a better fit for the rotating, throwaway nature of intermediate checkpoints than a model repository. Only changed files (and, within them, changed chunks) are transferred on each save.

Reviewer: ```suggestion
Set `push_to_bucket=True` to sync checkpoints to a [Hugging Face bucket](https://huggingface.co/docs/huggingface_hub/guides/buckets). Checkpoints still save to local disk, but the bucket receives a copy of each one. A bucket is mutable, Xet-backed object storage with no git history. Buckets suit intermediate checkpoints better than a model repository, since checkpoints rotate constantly and are meant to be thrown away. Each save transfers only the changed files, and only the changed chunks within them.
```

```py
from transformers import Trainer, TrainingArguments

trainer = Trainer(
    model=model,
    args=TrainingArguments(
        push_to_bucket=True,

>         bucket_id="my-org/my-run",

Reviewer 2: we could have a good default for this IMO, the least amount of args the better!

Author: Yeah, I didn't precise but it will default to `hub_model_id` if it is set or default to `output_dir` otherwise.

Author: so this is actually optional

        ...
    ),
    train_dataset=train_dataset,
)
trainer.train()
```

> Each checkpoint is uploaded to the bucket under its own `checkpoint-<step>/` prefix and **kept there even after it is rotated away locally** by [`~TrainingArguments.save_total_limit`] (i.e. the bucket accumulates every checkpoint). Uploads run asynchronously, so they don't block training.

Reviewer: ```suggestion
[`Trainer`] uploads each checkpoint to the bucket under its own `checkpoint-<step>/` prefix and keeps it there even after [`~TrainingArguments.save_total_limit`] rotates it off local disk.. The bucket accumulates every checkpoint. Uploads run asynchronously, so they don't block training.
```

> `push_to_bucket` is independent of `push_to_hub`: use it on its own to store checkpoints without creating a model repository, or together with `push_to_hub` to also publish the model. 

Reviewer: ```suggestion
`push_to_bucket` is independent of `push_to_hub`. Use it on its own to store checkpoints without creating a model repository, or together with `push_to_hub` to also publish the model. 
```

> To resume from a bucket, pass its handle to `resume_from_checkpoint`. [`Trainer`] downloads the checkpoint into `output_dir` and continues from it — handy for resuming on a fresh machine where nothing is on local disk yet. Pass the bucket root to resume from its latest checkpoint, or a full checkpoint handle to resume from a specific one:

Reviewer: ```suggestion
To resume from a bucket, pass its handle to `resume_from_checkpoint`. [`Trainer`] downloads the checkpoint into `output_dir` and continues from it. This helps when resuming on a fresh machine where nothing is on local disk yet. Pass the bucket root to resume from its latest checkpoint, or a full checkpoint handle to resume from a specific one:
```
